hazelcast/src/main/java/com/hazelcast/nio/tcp/TcpIpConnectionManager.java

```diff
@@ -193,25 +202,77 @@ private synchronized boolean bind(TcpIpConnection connection, Address remoteEndP
         if (logger.isFinestEnabled()) {
             logger.finest("Binding " + connection + " to " + remoteEndPoint + ", reply is " + reply);
         }
+
         final Address thisAddress = ioService.getThisAddress();
-        if (ioService.isSocketBindAny() && !connection.isClient() && !thisAddress.equals(localEndpoint)) {
-            String msg = "Wrong bind request from " + remoteEndPoint
-                    + "! This node is not requested endpoint: " + localEndpoint;
-            logger.warning(msg);
-            connection.close(msg, null);
+
+        // Some simple spoofing attack prevention
+        // Prevent BINDs from src that doesn't match the BIND local address
+        // Prevent BINDs from src that match us (same host & port)
+        if (SPOOFING_CHECKS
+                && (ensureValidBindSource(connection, remoteEndPoint)
+                 || ensureBindNotFromSelf(connection, remoteEndPoint, thisAddress))) {
+            return false;
+        }
+
+        // Prevent BINDs that don't have this node as the destination
+        if (ensureValidBindTarget(connection, remoteEndPoint, localEndpoint, thisAddress)) {
             return false;
         }
+
         connection.setEndPoint(remoteEndPoint);
         ioService.onSuccessfulConnection(remoteEndPoint);
         if (reply) {
             sendBindRequest(connection, remoteEndPoint, false);
         }
+
         if (checkAlreadyConnected(connection, remoteEndPoint)) {
             return false;
         }
+
         return registerConnection(remoteEndPoint, connection);
     }
 
+    private boolean ensureValidBindSource(TcpIpConnection connection, Address remoteEndPoint) {
+        try {
+            InetAddress originalRemoteAddr = connection.getRemoteSocketAddress().getAddress();
+            InetAddress presentedRemoteAddr = remoteEndPoint.getInetAddress();
+            if (!originalRemoteAddr.equals(presentedRemoteAddr)) {
```

Comment:
Yes I also considered that too, but I could come up with a setup that would violate this. Maybe my networking is weak. I considered the following:

2 Node on a private network behind NAT, each.
Trying to connect to each other, they should provide their public address, and **different** port, so this equality should still not pass.
Can you think of a case this would fail ? The kill-switch would be useful in these case, but even better if we can work around them anyhow.